File: proposed-projects/round-trip-experiment/reverse/20260909T020947Z_ffb57550/20260910T054605Z_51904c27/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

from .get_input_as_string import get_length_in_code_points

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]
    lines = src.split("\n")
    src_length = get_length_in_code_points(src)

    line_idx = 0
    while (
        line_idx < len(lines)
        and lines[line_idx]
        and pos > get_length_in_code_points(lines[line_idx]) - 1
        and pos < src_length
    ):
        pos -= get_length_in_code_points(lines[line_idx])
        line_idx += 1

    lines_to_show = lines[
        max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)) : line_idx + 1
    ]

    return [*lines_to_show, f"{' ' * max(0, pos)}^"]

File: packages/gbnf-experiment/data/20260909T020947Z_ffb57550/ported_implementation/src/utils/errors/build-error-position.ts
import { getLengthInCodePoints } from './get-input-as-string.js';

export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }
  const lines = src.split('\n');
  const srcLength = getLengthInCodePoints(src);

  let lineIdx = 0;
  while (
    lines[lineIdx] &&
    pos > getLengthInCodePoints(lines[lineIdx]) - 1 &&
    pos < srcLength
  ) {
    pos -= getLengthInCodePoints(lines[lineIdx]);
    lineIdx += 1;
  }

  const linesToShow = lines.slice(
    Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)),
    lineIdx + 1,
  );

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};
